### PR TITLE
ENH: port scipy.signal._arraytools to be Array API compatible

### PR DESCRIPTION
**Pull request by IvanYashchuk**
List of ported functions:
* odd_ext
* even_ext
* const_ext
* zero_ext

List of functions that were already working with Array API arrays:
* axis_slice
* axis_reverse

#### Reference issue

Ref. https://github.com/scipy/scipy/issues/15354

#### What does this implement/fix?

This PR makes functions from `scipy.signal._arraytools` accept and return arrays from Array API compatible libraries such as `numpy.array_api` and `cupy.array_api`. Only `numpy.array_api` is tested here.

cc: @AnirudhDagar, @rgommers, @czgdp1807

### CODE REVIEW

File: scipy/_lib/_array_api_util.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            f"Expected a single common namespace for array inputs, \
+              but got: {[n.__name__ for n in namespaces]}"
+        )
+
+    (xp,) = namespaces
+
+    return xp
+
+
+def _concatenate(arrays, axis):
```
Comment (czgdp1807) [inline scipy/_lib/_array_api_util.py:58]: I like this pattern. Creating a wrapper and using it as a drop in replacement across the library. It should be quite easy to deprecate in the distant future when `numpy.array_api` will stabilise.

**Comment by tupui** [comment]
Thanks @IvanYashchuk for the PR. Can you send an email to the mailing list linking it? Since this introduces the helper function, we should make all maintainers aware.

File: scipy/_lib/_array_api_util.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import numpy as np
+
+from scipy._lib import _pep440
+
+_NUMPY_INCLUDES_ARRAY_API = _pep440.parse(np.__version__) >= _pep440.Version(
+    "1.22.0"
+)
+
+
+def _get_namespace(*arrays):
```
Comment (tupui) [inline scipy/_lib/_array_api_util.py:19]: Personally I am not a fan of `get` something. I would prefer something even more explicit like `namespace_from_array` or something with array API in the name. Cf other comment on `_concatenate` about being explicit.

File: scipy/_lib/_array_api_util.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    # we will use assume the default (NumPy) namespace.
+    namespaces = {
+        x.__array_namespace__() if hasattr(x, "__array_namespace__") else np
+        for x in arrays
+    }
+
+    if len(namespaces) != 1:
+        raise ValueError(
+            f"Expected a single common namespace for array inputs, \
+              but got: {[n.__name__ for n in namespaces]}"
```
Comment (tupui) [inline scipy/_lib/_array_api_util.py:51]: This or we call unique and only return single occurrences of namespaces. This is to mitigate cases where user would through a large number of arrays in some functions.

File: scipy/_lib/_array_api_util.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    """
+    # `arrays` contains one or more arrays
+    # When there is no __array_namespace__,
+    # we will use assume the default (NumPy) namespace.
+    namespaces = {
+        x.__array_namespace__() if hasattr(x, "__array_namespace__") else np
+        for x in arrays
+    }
+
+    if len(namespaces) != 1:
```
Comment (tupui) [inline scipy/_lib/_array_api_util.py:48]: I agree here. Better to fail explicitly than trying out an arbitrary conversion from one backend to another. Otherwise we will have reports of weird behaviours and performances.

File: scipy/_lib/_array_api_util.py (hunk 1 above)
Comment (tupui) [reply scipy/_lib/_array_api_util.py:59]: Yes, but maybe it would be good to follow a common naming scheme to facilitate this and be clearer about where this is coming from.

**Review by tupui — COMMENTED** [review]
Looks easy to add, we just need to have these common functions like `_concatenate` then. A few comments for me.

File: scipy/_lib/_array_api_util.py (hunk 3 above)
Comment (czgdp1807) [reply scipy/_lib/_array_api_util.py:51]: > This is to mitigate cases where user would through a large number of arrays in some functions.

I was wondering the same. May be we can do something like below,

1. Obtain namespace of first array and assume that it is the answer.
2. Then from second array onwards, if namespace doesn't change then keep moving to thrid, fourth, fifth and so on.
3. Once we encounter a different namespace raise the error.

This wouldn't require usage of sets but it won't be verbose because, we will be exiting before we know the namespaces of all the arrays so we won't be able to show all the namespaces in our error message.

File: scipy/_lib/_array_api_util.py (hunk 2 above)
Comment (czgdp1807) [reply scipy/_lib/_array_api_util.py:19]: > `namespace_from_array`

+1. A better name.

**Comment by ev-br** [comment]
It'd be nice to understand the relation of this array api work to a parallel work which deals with the uarray backends. Are these two related? Complementary? Competing? 

(It's probably just my ignorance, but nonetheless)

File: scipy/_lib/_array_api_util.py (hunk 3 above)
Comment (IvanYashchuk) [reply scipy/_lib/_array_api_util.py:51]: `namespaces` is a set so only a single occurrence of each namespace is printed in the error message.

File: scipy/_lib/_array_api_util.py (hunk 2 above)
Comment (IvanYashchuk) [reply scipy/_lib/_array_api_util.py:19]: I'm fine with dropping the underscores, it's just what I noticed is used in SciPy, for example [`scipy._lib._util_asarray_validated`](https://github.com/scipy/scipy/blob/93939bb3dfeeb2a99ea23bba361fbf6df771155b/scipy/_lib/_util.py#L241).

I don't have a preference for the name of this function. Array API is mentioned in the module name `scipy._lib._array_api_util`. I can change it to `scipy._lib._array_api_util._namespace_from_arrays`.

**Comment by IvanYashchuk** [comment]
> It'd be nice to understand the relation of this array api work to a parallel work which deals with the uarray backends. Are these two related? Complementary? Competing?
> 
> (It's probably just my ignorance, but nonetheless)

Hey @ev-br! They are complementary. The goal is the same: make SciPy functions work with non-NumPy arrays (CuPy, Dask, PyTorch, JAX, etc.). The means to achieve that goal are different.

SciPy functions implemented with pure Python+NumPy functions can be ported to use Array API compatible modules, where we replace `numpy` module in SciPy with an `input_array.__array_namespace__()`. `__array_namespace__` carries a reference to a specific array library module it was created with. So we determine the library to be used dynamically based on the inputs of the function.

uarray backends are needed for the functions which directly rely on the memory buffers of NumPy arrays and use compiled code extensions. Here we need to be able to redirect SciPy function calls to an array library specific implementation (e.g. [`scipy.linalg.lu_factor`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lu_factor.html#scipy.linalg.lu_factor) -> [`cupyx.scipy.linalg.lu_factor`](https://docs.cupy.dev/en/stable/reference/generated/cupyx.scipy.linalg.lu_factor.html)).

File: scipy/_lib/_array_api_util.py (hunk 3 above)
Comment (tupui) [reply scipy/_lib/_array_api_util.py:53]: Great then!

File: scipy/_lib/_array_api_util.py (hunk 2 above)
Comment (tupui) [reply scipy/_lib/_array_api_util.py:19]: For the underscore we are a bit inconsistent yes. Let's see what others think.

File: scipy/_lib/_array_api_util.py (hunk 4 above)
Comment (tupui) [reply scipy/_lib/_array_api_util.py:50]: One thing we can consider, trying to convert lists into either np or the only other backend which is present.

Comment (czgdp1807) [reply scipy/_lib/_array_api_util.py:50]: That might lead to confusions, because the code will work and the user would think that they are receiving the optimum performance but that might not be the case as always. IMO, failing is the best option in case of conflicting namespaces.

File: scipy/signal/_arraytools.py [hunk 5]
```diff
@@ -232,10 +228,11 @@ def zero_ext(x, n, axis=-1):
...
            [ 0,  0,  0,  1,  4,  9, 16,  0,  0]])
     """
+    xp = _get_namespace(x)
     if n < 1:
         return x
     zeros_shape = list(x.shape)
     zeros_shape[axis] = n
-    zeros = np.zeros(zeros_shape, dtype=x.dtype)
-    ext = np.concatenate((zeros, x, zeros), axis=axis)
+    zeros = xp.zeros(zeros_shape, dtype=x.dtype)
```
Comment (czgdp1807) [inline scipy/signal/_arraytools.py:236]: This is supposed to fail, right, if `xp` doesn't have `zeros` attribute/method/function?

File: scipy/_lib/_array_api_util.py (hunk 4 above)
Comment (tupui) [reply scipy/_lib/_array_api_util.py:50]: The issue is that it would introduce a backward incompatibility in the sense that some user expect the function to work when they pass a list instead of an array of some type.

We could also add a (deprecation) warning instead of failing in that case so user can migrate their code.

Comment (czgdp1807) [reply scipy/_lib/_array_api_util.py:50]: > We could also add a (deprecation) warning instead of failing in that case so user can migrate their code.

Yes. Though, eventually we should fail in case of conflicting namespaces. For migrating to such failures, as you said, we can use `DeprecationWarning` for a better user experience.

File: scipy/signal/_arraytools.py (hunk 5 above)
Comment (IvanYashchuk) [reply scipy/signal/_arraytools.py:236]: If a library is Array API compliant then it must have the `zeros` function, https://data-apis.org/array-api/latest/API_specification/generated/signatures.creation_functions.zeros.html

File: scipy/_lib/_array_api_util.py (hunk 1 above)
Comment (tylerjereddy) [reply scipy/_lib/_array_api_util.py:61]: I suppose we should try to follow a similar path to `scikit-learn`. There's a long discussion here that Pamphile linked from the issue: https://github.com/scikit-learn/scikit-learn/issues/22352

Are we converging to similar approaches now?

Comment (tylerjereddy) [reply scipy/_lib/_array_api_util.py:61]: I also saw `scikit-learn` attempt to use NEP37 like https://github.com/scikit-learn/scikit-learn/pull/16574. It is hard to keep track of all the approaches. There's also NEP47 approach used here of course: https://numpy.org/neps/nep-0047-array-api-standard.html#appendix-a-possible-get-namespace-implementation

So, it is pretty much decided to drop/not accept NEP37 approach then? I initially went down that path because I had noticed that CuPy already had https://docs.cupy.dev/en/stable/reference/generated/cupy.get_array_module.html.

**Review by tylerjereddy — COMMENTED** [review]
Just a thought--is it easy to demonstrate a performance improvement with CuPy here? These seem like mostly private functions I'm less familiar with? I realize that demonstrating a performance improvement may be somewhat orthogonal, but on the other hand certainly can make things more compelling and help make sure that we don't i.e., lose out on practical performance gains on reasonable data sizes with HtoD and DtoH transfers for example.

I think typically you need fairly large arrays for gains--happy to test on a suitable node/GPU if there's a reproducer for a speedup here.

**Comment by IvanYashchuk** [comment]
The original idea was to start with a small PR to demonstrate the way for Array API compatibility.

I will prepare a demo showing performance improvement with CuPy. It shouldn't be difficult.

**Comment by tupui** [comment]
> Just a thought--is it easy to demonstrate a performance improvement with CuPy here? These seem like mostly private functions I'm less familiar with? I realize that demonstrating a performance improvement may be somewhat orthogonal, but on the other hand certainly can make things more compelling and help make sure that we don't i.e., lose out on practical performance gains on reasonable data sizes with HtoD and DtoH transfers for example.
> 
> I think typically you need fairly large arrays for gains--happy to test on a suitable node/GPU if there's a reproducer for a speedup here.

@mdhaber had a small experiment here https://github.com/mdhaber/scipy/pull/63

@rgommers do we have a concrete plan/timeline for this work?

**Comment by rgommers** [comment]
Thanks for circling back to this Tyler & all. Good timing, I was just writing up a larger proposal and will post it within the next 24 hours. 

tl;dr we do want to go ahead with array API standard support, and we have bandwidth to push it. We've learned some lessons from the more extensive efforts in scikit-learn on how to go about it. There's clearly going to be large performance gains that are achievable with GPU support; @mdhaber did some experiments with `scipy.stats` functionality on that I believe. Also read the PR description of https://github.com/scikit-learn/scikit-learn/pull/25956, that's pretty convincing:)

**Comment by rgommers** [comment]
> Good timing, I was just writing up a larger proposal and will post it within the next 24 hours.

Done in gh-18286. Now that the dust has settled and scikit-learn folks seem happy with the current incarnation, I'm much more comfortable with pushing this forward pretty quickly now.

**Comment by lucascolley** [comment]
superseded by gh-19001
